batched/system.go

```diff
@@ -0,0 +1,372 @@
+	s := &BatchProvidingSystem{
+		reprovideInterval: time.Hour * 24,
+		rsys:              provider,
+		keyProvider:       nil,
+		q:                 q,
+		ds:                datastore.NewMapDatastore(),
+		dynamicCh:         make(chan cid.Cid, 1),
+	}
+
+	for _, o := range opts {
+		if err := o(s); err != nil {
+			return nil, err
+		}
+	}
+
+	// This is after the options processing so we do not have to worry about leaking a context if there is an
+	// initialization error processing the options
+	ctx, cancel := context.WithCancel(context.Background())
+	s.ctx = ctx
+	s.close = cancel
+
+	return s, nil
+}
+
+func Datastore(batching datastore.Batching) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.ds = batching
+		return nil
+	}
+}
+
+func ReproviderInterval(duration time.Duration) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.reprovideInterval = duration
+		return nil
+	}
+}
+
+func KeyProvider(fn simple.KeyChanFunc) Option {
+	return func(system *BatchProvidingSystem) error {
+		system.keyProvider = fn
+		return nil
+	}
+}
+
+func (s *BatchProvidingSystem) Run() {
+	const pauseDetectionThreshold = time.Millisecond * 500
+	const maxCollectionDuration = time.Minute * 10
+	const maxTimeWaitingForProvides = time.Hour
+
+	provCh := s.q.Dequeue()
+
+	go func() {
+		s.closewg.Add(1)
+		defer s.closewg.Done()
+
+		m := make(map[cid.Cid]struct{})
+
+		maxCollectionDurationTimer := time.NewTimer(maxCollectionDuration)
+		pauseDetectTimer := time.NewTimer(maxTimeWaitingForProvides)
+		defer maxCollectionDurationTimer.Stop()
+		defer pauseDetectTimer.Stop()
+
+		for {
+			pauseDetectTimer.Reset(maxTimeWaitingForProvides)
+			maxCollectionDurationTimer.Reset(maxCollectionDuration)
+
+			performedReprovide := false
+		loop:
+			for {
+				select {
+				case <-maxCollectionDurationTimer.C:
+					emptyTimer(pauseDetectTimer)
+					break loop
+				default:
+				}
+
+				select {
+				case c := <-provCh:
+					m[c] = struct{}{}
+					resetTimer(pauseDetectTimer, pauseDetectionThreshold)
+					continue
+				default:
+				}
+
+				select {
+				case c := <-provCh:
+					m[c] = struct{}{}
+					resetTimer(pauseDetectTimer, pauseDetectionThreshold)
+				case c := <-s.dynamicCh:
+					m[c] = struct{}{}
+					resetTimer(pauseDetectTimer, pauseDetectionThreshold)
+					performedReprovide = true
+				case <-pauseDetectTimer.C:
+					emptyTimer(maxCollectionDurationTimer)
+					break loop
+				case <-maxCollectionDurationTimer.C:
+					emptyTimer(pauseDetectTimer)
+					break loop
+				case <-s.ctx.Done():
+					return
+				}
+			}
+
+			if len(m) == 0 {
+				continue
+			}
+
+			keys := make([]multihash.Multihash, 0, len(m))
+			for c := range m {
+				// hash security
+				if err := verifcid.ValidateCid(c); err != nil {
+					log.Errorf("insecure hash in reprovider, %s (%s)", c, err)
+					continue
+				}
+
+				keys = append(keys, c.Hash())
+				delete(m, c)
+			}
+
+			for !s.rsys.Ready() {
```

Comment:
do this _before_ deleting?

Reply:
Does it matter, we're going to sit in this for loop until either we close down the provider system or the system becomes ready.

Note: this means the providing + reproviding goroutines will end up blocking if the system is never ready. Calls to `Reprovide()` will block as well which seems reasonably in line with what the simple provider does (although that one waits for the reproviding to complete and this one only waits for them to be queued up).

Reply:
Oh, nvm. I thought we were continuing our outer loop for some reason.